### ISSUE TITLE
sh4div_test: inconsistency between WashingtonDC and Dreamcast

### ISSUE THREAD

**Issue by snickerbockers** [washingtondc-emu/washingtondc#26]
I've just gotten started with hardware testing over in washingtondc-emu/dctest@031c3af2ea3da27b714cb2ac24db7e647b79a9f9 and I've already caught a bug.  When I run sh4div_test, I get different output on the signed_div_test_16_16.

Also, the parts of the output that say SUCCESS are irrelevent; that only proves that my custom JIT'd division implementations are consistent with the ones gcc spits out.  I haven't come up with a good scheme for verifying that the division results are correct yet.

On WashingtonDC:
```
--
KallistiOS Git revision 374a47b: Sun Oct 15 20:39:53 PDT 2017
  jay@sbckrs_desktop:/home/jay/src/KallistiOS
maple: active drivers:
    Dreameye (Camera): Camera
    Sound Input Peripheral: Microphone
    PuruPuru (Vibration) Pack: JumpPack
    VMU Driver: Clock, LCD, MemoryCard
    Mouse Driver: Mouse
    Keyboard Driver: Keyboard
    Controller Driver: Controller
    Lightgun: LightGun
  DMA Buffer at ac0d99c0
vid_set_mode: 640x480IL NTSC
maple: attached devices:
  A0: Dreamcast Controller          (01000000: Controller)
attempting $d trials
==== unsigned_div_test_32_16 ====
1481765933 / 36047
the expected result is 41106
the actual result is 41106
SUCCESS
812669700 / 23988
the expected result is 33878
the actual result is 33878
SUCCESS
445349752 / 23000
the expected result is 19363
the actual result is 19363
SUCCESS
730417256 / 22183
the expected result is 32926
the actual result is 32926
SUCCESS
147699711 / 54335
the expected result is 2718
the actual result is 2718
SUCCESS
1889772843 / 47857
the expected result is 39487
the actual result is 39487
SUCCESS
777206980 / 24495
the expected result is 31729
the actual result is 31729
SUCCESS
1843638549 / 34344
the expected result is 53681
the actual result is 53681
SUCCESS
==== signed_div_test_16_16 ====
-4294983772 / 20713
the expected result is 0
the actual result is 0
SUCCESS
5719 / 12251
the expected result is 0
the actual result is 0
SUCCESS
25182 / 20512
the expected result is 1
the actual result is 1
SUCCESS
21755 / -4294983624
the expected result is -4294967297
the actual result is -4294967297
SUCCESS
-4294978136 / 78
the expected result is -4294967434
the actual result is -4294967434
SUCCESS
-4294998106 / -4294981868
the expected result is 2
the actual result is 2
SUCCESS
-4294996589 / -4294977755
the expected result is 2
the actual result is 2
SUCCESS
-4294994858 / -4294998492
the expected result is 0
the actual result is 0
SUCCESS
16 successes out of 16 total trials
arch: shutting down kernel
maple: final stats -- device count = 1, vbl_cntr = 45, dma_cntr = 45
vid_set_mode: 640x480IL NTSC
```

On a real Dreamcast:
```
--
KallistiOS Git revision 374a47b: Sun Oct 15 20:39:53 PDT 2017
  jay@sbckrs_desktop:/home/jay/src/KallistiOS
maple: active drivers:
    Dreameye (Camera): Camera
    Sound Input Peripheral: Microphone
    PuruPuru (Vibration) Pack: JumpPack
    VMU Driver: Clock, LCD, MemoryCard
    Mouse Driver: Mouse
    Keyboard Driver: Keyboard
    Controller Driver: Controller
    Lightgun: LightGun
  DMA Buffer at ac0d99c0
vid_set_mode: 640x480 NTSC
dc-load console support enabled
maple: attached devices:
  A0: Dreamcast Controller          (01000000: Controller)
  A1: Visual Memory                 (0e000000: Clock, LCD, MemoryCard)
  A2: Puru Puru Pack                (00010000: JumpPack)
attempting $d trials
==== unsigned_div_test_32_16 ====
1481765933 / 36047
the expected result is 41106
the actual result is 41106
SUCCESS
812669700 / 23988
the expected result is 33878
the actual result is 33878
SUCCESS
445349752 / 23000
the expected result is 19363
the actual result is 19363
SUCCESS
730417256 / 22183
the expected result is 32926
the actual result is 32926
SUCCESS
147699711 / 54335
the expected result is 2718
the actual result is 2718
SUCCESS
1889772843 / 47857
the expected result is 39487
the actual result is 39487
SUCCESS
777206980 / 24495
the expected result is 31729
the actual result is 31729
SUCCESS
1843638549 / 34344
the expected result is 53681
the actual result is 53681
SUCCESS
==== signed_div_test_16_16 ====
-16476 / 20713
the expected result is 0
the actual result is 0
SUCCESS
5719 / 12251
the expected result is 0
the actual result is 0
SUCCESS
25182 / 20512
the expected result is 1
the actual result is 1
SUCCESS
21755 / -16328
the expected result is -1
the actual result is -1
SUCCESS
-10840 / 78
the expected result is -138
the actual result is -138
SUCCESS
-30810 / -14572
the expected result is 2
the actual result is 2
SUCCESS
-29293 / -10459
the expected result is 2
the actual result is 2
SUCCESS
-27562 / -31196
the expected result is 0
the actual result is 0
SUCCESS
16 successes out of 16 total trials
arch: shutting down kernel
maple: final stats -- device count = 3, vbl_cntr = 9, dma_cntr = 9
vid_set_mode: 640x480 NTSC
```

**Comment by snickerbockers** [comment]
I hardcoded the dividend to 4294950820 and the divisor to 20713 (this is what they are during the first iteration according to gdb).  The output was different here too; whatever this bug is, it effects the behavior of both the rand function *and* printf.

On WashingtonDC:
```
-4294983772 / 20713
the expected result is 0
the actual result is 0
SUCCESS
```

On Dreamcast:
```
-16476 / 20713
the expected result is 0
the actual result is 0
```

**Comment by snickerbockers** [comment]
Actually, the discrepancy is entirely contained to printf.  When I change replace every %d with %x, they both output the exact same results.   I've never had this sort of problem with printf before, but I generally don't dump rand() into it either; is it possible that the discrepancy is caused by really big ints?

**Comment by snickerbockers** [comment]
If I printf INT_MAX, both Dreamcast and WashingtonDC will print 2147483647; this is the correct value.  

If I printf INT_MAX+1, Dreamcast outputs -2147483648, but WashingtonDC flips the fuck out and prints -6442450944.

So *maybe* something to do with negative numbers?  Or maybe only really big negative numbers?  It must be relevant that it occurs right at the point of overflow...

**Comment by snickerbockers** [comment]
The problem appears to be somewhere in the vicinity of PC=0x8c075fa0.  the _uquad variable in _vfprintf_r (see KallistiOS' source) is a 64-bit int which occupies r13 and r14, with r14 being the most-significant-word.  This variable represents the unsigned magnitude of the value being printed (the hyphen character for negative numbers is handled separately).

When you print INT_MAX+1 (0x80000000) on WashingtonDC, the output is -6442450944, as stated above.  What I'm seeing in this case is that r13 is the correct value (0x80000000), but r14 is 1; this is actually what makes _uquad 6442450944.  What I need to do now is figure out why that one bit gets set in r14.  I think it's safe to assume that it is not supposed to be set.

**Comment by snickerbockers** [comment]
Maybe negc isn't setting the T-flag (borrow, in this context) correctly...

**Comment by snickerbockers** [comment]
I came up with the following test, which proves that there is a discrepancy between how Dreamcast sets the T flag from within the NEGC instruction, and how WashingtonDC does it when the input is INT_MAX + 1.

(don't mind the fact that  the input is seemingly different, that's a result of the same printf bug I'm trying to fix here).

WashingtonDC:
```
the input is -6442450944
about to call 0x8c0b140c
the output is 0x40000000
```

Dreamcast:
```
Just a test: -2147483648
the input is -2147483648
about to call 0x8c0b140c
the output is 0x40000001
```

Source:
```
static int test_negc(int input) {                                                                                    
    char const *prog_asm =                                                                                           
        "clrt\n"                                                                                                     
        "negc r4, r4\n"                                                                                              
        "rts\n"                                                                                                      
        "stc sr, r0\n";                                                                                              
    printf("the input is %d\n", input);                                                                              
    clear_jit();                                                                                                     
    sh4asm_input_string(prog_asm);                                                                                   
    refresh_inst_list();                                                                                             
    test_fn func_ptr = (test_fn)inst_list;                                                                           
    printf("about to call %p\n", func_ptr);                                                                          
    printf("the output is 0x%08x\n", func_ptr(input, 0));                                                            
}                                                                                                                    
```

**Comment by snickerbockers** [comment]
Ugh, that last commit didn't quite fix *everything*.  The initial printf at the top is still wrong even though almost everything else is right...

**Comment by snickerbockers** [comment]
This issue has been resolved by 248334a1d82a6c4c4f936f345952332dbae2eaa9 and 29f07dfeacc4da8961d25ac9840851af0ccf1a4a.
